### PR TITLE
Add zero-width space before the last filename dot to avoid splitting the extension

### PR DESCRIPTION
**Pull request by squelart**
#### 07f25ac1ae5727fa1bf2e242ed57e3a4efe3306a
<pre>
Add zero-width space before the last filename dot to avoid splitting the extension
<a href="https://bugs.webkit.org/show_bug.cgi?id=255390">https://bugs.webkit.org/show_bug.cgi?id=255390</a>
rdar://99658593

Reviewed by Cameron McCormack.

This patch also fixes an issue where the LTR isolation characters were missing around updated filenames in wide-layout attachments; added relevant test.

* LayoutTests/fast/attachment/attachment-title-with-rtl-expected.html:
* LayoutTests/fast/attachment/attachment-title-with-rtl.html:
Tweaked test text to avoid wrapping in 2nd attachemnt; it caused sub-pixels differences, and it&apos;s not relevant to this test.
* LayoutTests/fast/attachment/mac/wide-attachment-title-with-rtl-expected.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-title-with-rtl.html: Added.
Copy of non-wide tests, updated to exercise wrapping in 3rd attachment.
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
Updated to account for added character.
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::attributeChanged):
Actually use attachmentTitleForDisplay when the title changes. Also with action and subtitles for consistency and future-proofing.
(WebCore::HTMLAttachmentElement::attachmentTitleForDisplay const):
Add zero-width spaces before dot and file extension.

Canonical link: <a href="https://commits.webkit.org/263045@main">https://commits.webkit.org/263045@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/48620bbd41031ec95631980fd817d6e1cbf303fe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3327 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3386 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3502 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4744 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3657 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3304 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3449 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3409 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2889 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3367 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3652 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3001 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4566 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1155 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2962 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2880 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2937 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3016 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4312 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3400 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2720 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2962 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2963 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/836 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2964 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3238 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/html/HTMLAttachmentElement.cpp [hunk 1]
```diff
@@ -469,13 +469,13 @@ void HTMLAttachmentElement::attributeChanged(const QualifiedName& name, const At
 
     if (name == actionAttr) {
         if (m_actionTextElement)
-            m_actionTextElement->setTextContent(String(newValue.string()));
+            m_actionTextElement->setTextContent(String(attachmentActionForDisplay()));
```
Comment (heycam) [inline Source/WebCore/html/HTMLAttachmentElement.cpp:473]: Nit: I guess this is just a matter of style, but I think it's more common to see `someAtom.string()` rather than `String(someAtom)` in the codebase.

File: LayoutTests/fast/attachment/attachment-title-with-rtl.html [hunk 2]
```diff
@@ -1,5 +1,8 @@
 <!DOCTYPE html><!-- webkit-test-runner [ AttachmentElementEnabled=true ] -->
 <html>
+<head>
+<meta name="fuzzy" content="maxDifference=0-100; totalPixels=0-50" />
```
Comment (heycam) [inline LayoutTests/fast/attachment/attachment-title-with-rtl.html]: Hmm, why is this needed? Does the text layout differ on some platforms with these control characters inserted?

**Review by heycam — APPROVED** [review]

File: LayoutTests/fast/attachment/attachment-title-with-rtl.html (hunk 2 above)
Comment (squelart) [reply LayoutTests/fast/attachment/attachment-title-with-rtl.html]: Here's an example of error I got: https://ews-build.s3-us-west-2.amazonaws.com/macOS-Monterey-Release-WK2-Tests-EWS/6f82ec3e-2814/results.html (expand the first failure). You'll notice some very slightly different pixels around a couple of characters in the 2nd attachment.
I believe the only difference I introduced are those two zero-width spaces around the dot, which somehow must disturb the exact sub-pixel location near the start of the string. Should I file an issue for that?

File: Source/WebCore/html/HTMLAttachmentElement.cpp (hunk 1 above)
Comment (squelart) [reply Source/WebCore/html/HTMLAttachmentElement.cpp:473]: `AtomString::string()` returns a `const String&`, however `setTextContent` takes an rvalue-ref, so it doesn't work in this case; instead I must force the creation of a temporary `String`.

As a style choice, maybe we could revisit the extensive use of rvalue references, in favor of plain values? By-value parameters allow efficient moves while still allowing copies. More arguments there: https://google.github.io/styleguide/cppguide.html#Rvalue_references . Though I would guess the preference for rvalue references may be due to the fear of unchecked copying, and encouraging developers to find solutions with movable objects.

File: LayoutTests/fast/attachment/attachment-title-with-rtl.html (hunk 2 above)
Comment (squelart) [reply LayoutTests/fast/attachment/attachment-title-with-rtl.html]: As discussed, I'll wait until we have some understanding of what's happening, before landing this patch (in case there's a more appropriate fix for the test).

File: Source/WebCore/html/HTMLAttachmentElement.cpp (hunk 1 above)
Comment (heycam) [reply Source/WebCore/html/HTMLAttachmentElement.cpp:473]: > however setTextContent takes an rvalue-ref

OK, I didn't notice that.

**Comment by squelart** [comment]
Trying something with the update:
Only adding a zero-width space *before* the dot, so that the dot will preferably stay with the extension (if that extension doesn't fit on the line).
And I've removed the image-diff tolerance in the RTL test, to see if the above change helps.

File: LayoutTests/fast/attachment/attachment-title-with-rtl.html (hunk 2 above)
Comment (squelart) [reply LayoutTests/fast/attachment/attachment-title-with-rtl.html]: I've tweaked the test to avoid this issue (while still testing RTL characters).
